openjdk/mmtkVMCompanionThread.cpp

```diff
@@ -0,0 +1,118 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "mmtk.h"
+#include "mmtkVMCompanionThread.hpp"
+#include "runtime/mutex.hpp"
+
+MMTkVMCompanionThread::MMTkVMCompanionThread():
+    NamedThread(),
+    _desired_state(_threads_resumed),
+    _reached_state(_threads_resumed),
+    _resumption_count(0) {
+  set_name("MMTK VM Companion Thread");
+  _lock = new Monitor(Monitor::nonleaf,
+                      "MMTkVMCompanionThread::_lock",
+                      true,
+                      Monitor::_safepoint_check_never);
+}
+
+MMTkVMCompanionThread::~MMTkVMCompanionThread() {
+  guarantee(false, "MMTkVMCompanionThread deletion must fix the race with VM termination");
+}
+
+void MMTkVMCompanionThread::run() {
+  this->initialize_named_thread();
+
+  for (;;) {
+    // Wait for suspend request
+    log_trace(gc)("MMTkVMCompanionThread: Waiting for suspend request...");
+    {
+      MutexLockerEx locker(_lock, Mutex::_no_safepoint_check_flag);
+      assert(_reached_state == _threads_resumed, "Threads should be running at this moment.");
+      while (_desired_state != _threads_suspended) {
+        _lock->wait(true);
+      }
+      assert(_reached_state == _threads_resumed, "Threads should still be running at this moment.");
+    }
+
+    // Let the VM thread stop the world.
+    log_trace(gc)("MMTkVMCompanionThread: Letting VMThread execute VM op...");
+    VM_MMTkSTWOperation op(this);
+    VMThread::execute(&op);
+
+    // Tell the waiter thread that the world has resumed.
+    log_trace(gc)("MMTkVMCompanionThread: Notifying threads resumption...");
+    {
+      MutexLockerEx locker(_lock, Mutex::_no_safepoint_check_flag);
+      assert(_desired_state == _threads_resumed, "start-the-world should be requested.");
+      assert(_reached_state == _threads_suspended, "Threads should still be suspended at this moment.");
+      _reached_state = _threads_resumed;
+      _resumption_count++;
+      _lock->notify_all();
+    }
+  }
+}
+
+void MMTkVMCompanionThread::request(stw_state desired_state, bool wait_until_reached) {
+  assert(!Thread::current()->is_VM_thread(), "Requests can only be made by GC threads. Found VM thread.");
+  assert(Thread::current() != this, "Requests can only be made by GC threads. Found companion thread.");
+  assert(!Thread::current()->is_Java_thread(), "Requests can only be made by GC threads. Found Java thread.");
+
+  MutexLockerEx locker(_lock, Mutex::_no_safepoint_check_flag);
+  assert(_desired_state != desired_state, "State %d already requested.", desired_state);
+  _desired_state = desired_state;
+  _lock->notify_all();
+
+  if (wait_until_reached) {
+    while (_reached_state != desired_state) {
+      _lock->wait(true);
+    }
+  }
+}
+
+void MMTkVMCompanionThread::wait_for_reached(stw_state desired_state) {
```

Comment:
This seems not used anywhere.

Reply:
My intention was to allow two different ways to use the `request` method:

1.  Call `request(some_state, true)`. The thread will block until reaching `some_state`.
2.  Call `request(some_state, false)`, do something else, and then call `wait_for_reached(some_state)`.  This allows some things to be done while the VM is stopping threads.

Currently in `mmtkUpcalls.cpp`, `mmtk_stop_all_mutators` and `mmtk_resume_mutators` only uses the first way, because this is the interface mmtk-core requires the VM to implement.  But I think it may be helpful in the future to allow this usage.